flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1492,13 +1532,27 @@ private void checkpointStreamOperator(StreamOperator<?> op) throws Exception {
 
 		RecordWriter<SerializationDelegate<StreamRecord<OUT>>> output = new RecordWriterBuilder<SerializationDelegate<StreamRecord<OUT>>>()
 			.setChannelSelector(outputPartitioner)
-			.setTimeout(bufferTimeout)
-			.setTaskName(taskName)
+			.setFlushAlways(bufferTimeout == 0)
 			.build(bufferWriter);
 		output.setMetricGroup(environment.getMetricGroup().getIOMetricGroup());
+
+		if (bufferTimeout > 0) {
+			outputFlushers.addOutputFlusher(startOutputFlusher(output, taskName, bufferTimeout, mailboxExecutor));
+		}
 		return output;
 	}
 
+	private static OutputFlusher startOutputFlusher(
+			RecordWriter<?> output,
+			String taskName,
+			long bufferTimeout,
+			MailboxExecutor mailboxExecutor) {
+		Preconditions.checkArgument(bufferTimeout >= -1);
```

Comment:
This should be in `OutputFlusher` because it makes sense for any `OutputFlusher`.
Also, values -1, 0 and maybe under 10 don't make sense.

Reply:
True, here values -1 and 0 do not make sense, as now they are handled in a bit different way.
`-1` means output flushing disabled
`0` means flush always
in both cases, there is no `OutputFlusher`, so I've adjusted the `checkArgument` and moved it to the `OutputFlusher` constructor.

Values `(0, 10]` do make sense, and there are users requiring flushes that often. Actually a lot of effort was put into the code https://issues.apache.org/jira/browse/FLINK-8581 to make it work reasonably well with ~1ms `bufferTimeout` (and we are benchmarking those cases as well :) )